cli/mender-artifact/write.go

```diff
@@ -460,3 +471,125 @@ func extractKeyValues(params []string) (*map[string]string, error) {
 	}
 	return keyValues, nil
 }
+
+// SSH to remote host and dump rootfs snapshot to a local temporary file.
+func getDeviceSnapshot(host string) (string, error) {
+	var userAtHost string
+	port := "22"
+	if remotePort := strings.Split(host, ":"); len(remotePort) == 2 {
+		port = remotePort[1]
+		userAtHost = remotePort[0]
+	} else {
+		userAtHost = host
+	}
+
+	// First echo to stdout such that we know when ssh connection is
+	// established (password prompt is written to /dev/tty directly,
+	// and hence impossible to detect).
+	args := []string{"-p", port, userAtHost, "sudo", "-S", "/bin/sh", "-c",
+		"'echo \"aLL SET!\"; mender snapshot dump | cat'"}
+
+	cmd := exec.Command("ssh", args...)
+
+	// Simply connect stdin/stderr
+	cmd.Stdin = os.Stdin
+	cmd.Stderr = os.Stderr
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", errors.New("Error redirecting stdout on exec")
+	}
+
+	if err := cmd.Start(); err != nil {
+		return "", err
+	}
+
+	// Wait for 60 seconds for ssh to establish connection
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	errChan := make(chan error)
+	go func() {
+		stdoutRdr := bufio.NewReader(stdout)
+		for {
+			line, err := stdoutRdr.ReadString('\n')
+			if err != nil {
+				errChan <- errors.Wrap(err, "Error waiting for "+
+					"device to start dumping snapshot")
+				return
+			}
+			if strings.Contains(line, "aLL SET!") {
+				break
+			}
+		}
+		errChan <- nil
+	}()
+	select {
+	case err = <-errChan:
+		// Go routine finished
+
+	case <-ctx.Done():
+		// Context deadline exceeded
+		err = ctx.Err()
+	}
+
+	if err != nil {
+		cmd.Process.Kill()
+		return "", err
+	}
+	f, err := ioutil.TempFile("", "rootfs.ext4")
+	if err != nil {
+		cmd.Process.Kill()
+		return "", err
+	}
+	defer f.Close()
+	filePath := f.Name()
+
+	if cmd.ProcessState != nil && cmd.ProcessState.Exited() {
+		return "", errors.New("SSH session closed unexpectedly")
+	}
+
+	fmt.Print("Receiving snapshot from device...")
+	buf := make([]byte, 1024*1024*32)
+	var written int64
+	for {
+		nr, err := stdout.Read(buf)
+		if err != nil {
+			cmd.Process.Kill()
+			return "", errors.Wrap(err,
+				"Error receiving snapshot from device")
+		}
+		nw, err := f.Write(buf[:nr])
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			cmd.Process.Kill()
+			return "", errors.Wrap(err, "Error storing snapshot locally")
+		} else if nw < nr {
+			cmd.Process.Kill()
+			return "", io.ErrShortWrite
+		}
+		written += int64(nw)
+		fmt.Printf("\rReceiving snapshot from device... [%8s]", sizeStr(written))
```

Comment:
Will this be intermingled with the output from the `mender` command on the device? Because stderr is connected to directly, so I assume that output will make it across.

Reply:
I made it so that the progress bar will not print if the file descriptor is not a TTY device (in  `mender`). On a second thought I think it would be better to make it more explicit by adding a `--quiet` flag to the `snapshot dump` command.

Reply:
For some reason ssh trims `\r` characters from the stderr stream, that's why I disabled the progress bar for non-tty devices in general.